spring-web/src/main/java/org/springframework/http/client/reactive/JettyClientHttpConnector.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.http.client.reactive;
+
+import java.net.URI;
+import java.util.function.Function;
+
+import org.eclipse.jetty.client.HttpClient;
+import reactor.core.publisher.Mono;
+
+import org.springframework.context.SmartLifecycle;
+import org.springframework.core.io.buffer.DataBufferFactory;
+import org.springframework.core.io.buffer.DefaultDataBufferFactory;
+import org.springframework.http.HttpMethod;
+
+/**
+ * Jetty implementation of {@link ClientHttpConnector}.
+ *
+ * @author Sebastien Deleuze
+ * @since 5.1
+ */
+public class JettyClientHttpConnector implements ClientHttpConnector, SmartLifecycle {
+
+	private final HttpClient httpClient;
+
+	private DataBufferFactory bufferFactory = new DefaultDataBufferFactory();
+
+
+	/**
+	 * Create a Jetty {@link ClientHttpConnector} with the default {@link HttpClient}.
+	 */
+	public JettyClientHttpConnector() {
+		this(new HttpClient());
+	}
+
+	/**
+	 * Create a Jetty {@link ClientHttpConnector} with the given {@link HttpClient}.
+	 */
+	public JettyClientHttpConnector(HttpClient httpClient) {
+		this.httpClient = httpClient;
+	}
+
+
+	public void setBufferFactory(DataBufferFactory bufferFactory) {
+		this.bufferFactory = bufferFactory;
+	}
+
+	@Override
+	public Mono<ClientHttpResponse> connect(HttpMethod method, URI uri,
+			Function<? super ClientHttpRequest, Mono<Void>> requestCallback) {
+
+		if (!httpClient.isStarted()) {
+			try {
+				this.httpClient.start();
+			}
+			catch (Exception ex) {
+				return Mono.error(ex);
+			}
+		}
+
+		if (!uri.isAbsolute()) {
+			return Mono.error(new IllegalArgumentException("URI is not absolute: " + uri));
+		}
+
+		JettyClientHttpRequest clientHttpRequest =
+				new JettyClientHttpRequest(httpClient.newRequest(uri).method(method.toString()), bufferFactory);
+		return requestCallback.apply(clientHttpRequest).then(clientHttpRequest.getResponse());
+	}
+
+	@Override
+	public boolean isAutoStartup() {
+		return true;
+	}
+
+	@Override
+	public void stop(Runnable callback) {
+		stop();
+		callback.run();
+	}
+
+	@Override
+	public void start() {
+		try {
+			this.httpClient.start();
```

Comment:
A code level comment that explains HttpClient is internally synchronized and protected with state checks around here could be useful.